concurrency/atomicmap.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2024 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package concurrency
+
+import (
+	"sync"
+
+	"golang.org/x/exp/constraints"
+)
+
+type AtomicValue[T constraints.Integer] struct {
```

Comment:
For this struct, why not using `sync.Atomic` (like `atomic.Int32`) instead? That's lock-free and should have a lot better performance.

Reply:
That was the initial idea, but one of the way I'm using it now is with Int64, and another one as UInt32, so it needs to be flexible. We looked at using `atomic.Value`, but it doesn't implement `Add()`, which we need.

Reply:
I'd recommend creating 2 structs then, the use of locking is orders of magnitude slower than using atomic ([some source](https://stackoverflow.com/a/47446114/192024))

There's definitely an Add: https://pkg.go.dev/sync/atomic#Int32.Add

Reply:
That's a good point! I was curious and did a little benchmark of an atomic vs mutex add:

```
BenchmarkAtomicAdd-12    	21512379	        59.52 ns/op	       0 B/op	       0 allocs/op
BenchmarkMutexAdd-12     	10634985	       113.2 ns/op	       0 B/op	       0 allocs/op
```

So maybe not orders of magnitude, but the mutex-based one is almost twice as slow.

So I guess it's a question of "dirtying up" the code and getting the performance benefit, or keeping it cleaner/more maintainable, but paying the latency penalty.
I don't have a strong opinion here, to be honest, but I usually lean more towards maintainability, unless it's a really high-load system with operations per second in the hundreds of thousands. What do you all think?

Reply:
I spent some time exploring the alternative to workaround the generics to handle both cases with a single implementation but I got blocked by limited typecast with generics in Go (I cannot do `return (T) myInt64`, for example)

I do agree that duplicating the map seems overkill unless there is a justified use case that demands it - classic premature optimization case for now IMO.

This is me adding the Int64 variant of this map (for future reference)
https://github.com/artursouza/kit/commit/94fa3f4d82c1882f5801abcb4be26dcac7b4d4ae

Unless there are more data points, I recommend we merge this code with a single implementation of the atomic map.

Reply:
I agree, I think this is ok to merge as is. I think the usage of this package will not likely be a limiting factor for performance and the benchmark diff isn't worth the code sprawl.